File: src/hetzner_code/commands/add.py
from __future__ import annotations

from pathlib import Path

import click

from hetzner_code import attach, state
from hetzner_code.github import parse_repo_url
from hetzner_code.run import HcodeError


def add(
    *,
    repo_url: str,
    instance_name: str,
    branch: str | None,
    env_files: tuple[str, ...],
) -> None:
    instance = state.load(instance_name)
    repo = parse_repo_url(repo_url)

    if any(r.name == repo.name for r in instance.repos):
        raise HcodeError(f"'{repo.name}' is already on instance '{instance_name}'")

    if instance.repos:
        existing = ", ".join(r.name for r in instance.repos)
        click.echo(
            f"note: '{instance_name}' already runs [{existing}] — adding another codebase "
            "means they share this box's CPU. Your call; `hcode status` shows the type."
        )

    click.echo(f"attaching {repo.slug} to '{instance_name}' ({instance.ip}) ...")
    repo_state = attach.attach_repo(
        ip=instance.ip,
        identity_path=Path(instance.login_key_path),
        instance_name=instance_name,
        repo_url=repo_url,
        branch=branch,
        env_files=env_files,
    )
    instance.repos.append(repo_state)
    state.save(instance)

    click.echo(f"done: {repo.name} is on {instance_name} at /root/code/{repo.name}")
    click.echo(f"  ssh:  hcode ssh {instance_name} --repo {repo.name}")

File: internal/commands/add.go
package commands

import (
	"fmt"
	"strings"

	"github.com/dikee/hcode/internal/attach"
	"github.com/dikee/hcode/internal/github"
	"github.com/dikee/hcode/internal/run"
	"github.com/dikee/hcode/internal/state"
)

// AddOptions are every flag `hcode add` accepts.
type AddOptions struct {
	RepoURL      string
	InstanceName string
	Branch       string
	EnvFiles     []string
}

// Add runs `hcode add`.
func Add(opts AddOptions) error {
	instance, err := state.Load(opts.InstanceName)
	if err != nil {
		return err
	}
	repo, err := github.ParseRepoURL(opts.RepoURL)
	if err != nil {
		return err
	}

	for _, r := range instance.Repos {
		if r.Name == repo.Name {
			return run.Errorf("'%s' is already on instance '%s'", repo.Name, opts.InstanceName)
		}
	}

	if len(instance.Repos) > 0 {
		names := make([]string, len(instance.Repos))
		for i, r := range instance.Repos {
			names[i] = r.Name
		}
		fmt.Printf(
			"note: '%s' already runs [%s] — adding another codebase means they share this "+
				"box's CPU. Your call; `hcode status` shows the type.\n",
			opts.InstanceName, strings.Join(names, ", "),
		)
	}

	fmt.Printf("attaching %s to '%s' (%s) ...\n", repo.Slug(), opts.InstanceName, instance.IP)
	var branchPtr *string
	if opts.Branch != "" {
		branchPtr = &opts.Branch
	}
	repoState, err := attach.AttachRepo(instance.IP, instance.LoginKeyPath, opts.InstanceName, opts.RepoURL, branchPtr, opts.EnvFiles)
	if err != nil {
		return err
	}
	instance.Repos = append(instance.Repos, repoState)
	if err := state.Save(instance); err != nil {
		return err
	}

	fmt.Printf("done: %s is on %s at /root/code/%s\n", repo.Name, opts.InstanceName, repo.Name)
	fmt.Printf("  ssh:  hcode ssh %s --repo %s\n", opts.InstanceName, repo.Name)
	return nil
}
